pennylane/transforms/add_noise.py

```diff
@@ -0,0 +1,315 @@
+        TransformProgram(cancel_inverses, merge_rotations, undo_swaps, _expand_metric_tensor, batch_transform, expand_fn, add_noise)
+
+        Finally, more precise control over the insertion of the transform can be achieved by specifying
+        an integer or slice for indexing for extracting the transform program. For example, one can do:
+
+        >>> qml.transforms.add_noise(circuit, noise_model, level=2).transform_program
+        TransformProgram(cancel_inverses, merge_rotations, add_noise)
+        >>> qml.transforms.add_noise(circuit, noise_model, level=slice(1,3)).transform_program
+        TransformProgram(merge_rotations, undo_swaps, add_noise)
+
+    """
+    if not isinstance(noise_model, qml.NoiseModel):
+        raise ValueError(
+            f"Argument noise_model must be an instance of NoiseModel, got {noise_model}."
+        )
+
+    if level is None or level == "user":  # decompose templates and their adjoints
+
+        def stop_at(obj):
+            if not isinstance(obj, qml.operation.Operator):
+                return True
+            if not obj.has_decomposition:
+                return True
+            return not (hasattr(qml.templates, obj.name) or isinstance(obj, qml.ops.Adjoint))
+
+        error_type = (qml.operation.DecompositionUndefinedError,)
+        decompose = qml.devices.preprocess.decompose
+        [tape], _ = decompose(tape, stopping_condition=stop_at, name="add_noise", error=error_type)
+
+    conditions, noises = [], []
+    metadata = noise_model.metadata
+    for condition, noise in noise_model.model_map.items():
+        conditions.append(lru_cache(maxsize=512)(condition))
+        noises.append(qml.tape.make_qscript(noise))
+
+    new_operations = []
+    for operation in tape.operations:
+        curr_ops = [operation]
+        for condition, noise in zip(conditions, noises):
+            if condition(operation):
+                noise_ops = noise(operation, **metadata).operations
+                if operation in noise_ops and _check_queue_op(operation, noise, metadata):
+                    ops_indx = noise_ops.index(operation)
+                    curr_ops = noise_ops[:ops_indx] + curr_ops + noise_ops[ops_indx + 1 :]
+                else:
+                    curr_ops.extend(noise_ops)
+        new_operations.extend(curr_ops)
+
+    new_tape = type(tape)(new_operations, tape.measurements, shots=tape.shots)
+    post_processing_fn = qml.devices.preprocess.null_postprocessing
+
+    return [new_tape], post_processing_fn
+
+
+def _check_queue_op(operation, noise_func, metadata):
+    """Performs a secondary check for existence of an operation in the queue using a randomized ID"""
+
+    test_id = "f49968bfc4W0H86df3A733bf6e92904d21a_!$-T-@!_c131S549b169b061I25b85398bfd8ec1S3c"
+    test_queue = noise_func(
+        qml.noise.partial_wires(operation, id=test_id)(operation.wires), **metadata
+    ).operations
+
+    return test_id in [getattr(o, "id", "") for o in test_queue]
```

Comment:
(Not blocking)

I know we need this to support the required functionality, and its the most elegant thing we can do for now. However, this type of "double checking" is not clean or efficient and I think with better design can be avoidable. We should consider if it's possible to have something cleaner and more efficient in future PRs.

Reply:
I agree. I think we can iterate on improving this part of the functionality in future work.

For now, regarding the efficiency part, this will called only if the first check passes, which means someone is actually trying to use a custom sequencing of noise operations. As of now, I don't think that will happen that often. Currently, this is linear in the number of operations queued by a `noise_fn`. We could slightly improve it using a `generator` like approach via `any` or some kind of `index` pre-processing, but I don't think the gain will be significant.

Reply:
I guess it will run for each operation in the circuit that passes the condition right? So if I have n gates in my circuit and k of them pass the conditional, and I have m gates in the noise model, the run time is like O(m*k)

Reply:
Right! I can make this worst case `O(m*k)` (from the average case) by using `any`, will push it in the next commit.